README
This plugin requires you to be on the dev channel.
http://sites.google.com/a/chromium.org/dev/getting-involved/dev-channel

The UI currently sucks, it has been ripped off another extension. It should be fixed soon. 
To add torrents, right click on a link and choose add to utorrent.

To install this, visit chrome://extensions/ and select Load unpacked extension. Then select this folder.

changelog.txt
0.6.2
 * changed get torrents to download data incrementally

0.6.1
* Secondary server added for dyndns

0.6
* Ability to add torrents from any pane

0.5
* Basic display of torrents and manipulation


js/utorrent.js
function isset(variable)
{
    return (typeof variable != 'undefined');
}


function setupUtorrent(errorFn) {
    var server = {};
    $.each(["ip", "alt_ip", "port", "login", "password"], function(i, item){
        server[item] = localStorage["utorrent_" + item];
    })
    

    if (!(isset(server.ip)))
    {
        $notify.txt("Error!", "This extension has not been configured!")
        chrome.tabs.create({url:'chrome-extension://'+location.hostname+'/options.html'});        
        return {not_configured: true}
    }
    $.ajaxSetup({timeout: 5000,username: server.login, password: server.password});
    var baseUrl = function(ip){return "http://" + server[ip]+":"+server.port+"/gui/";};
    var cur_data;

    var withToken = function(fn) {
        if (isset(cur_data)) {
            fn.apply(this, cur_data);
        } 
        else {
            var xhrs = [];
            var success = function(item){
                return function(data) {
                    if (!(isset(cur_data))) {
                        cur_data = [$(data).text(), baseUrl(item)];
                        $.each(xhrs, function(i,x){x.abort();})
                        fn.apply(this, cur_data);
                    }
                }
            };
            var times = 0;
            var error = function() {
                times += 1;
                if (times >= 2) { 
                    return errorFn.apply(this, arguments);
                }
            };
            xhrs = $.map(["ip", "alt_ip"], function(item) {
                    return $.ajax({url: baseUrl(item) + "token.html", success: success(item), error: error})
            })
        }
    }
    
    var get = function(data, fn) {
        withToken(function(token, url) {
            $.ajax({url: url, data: $.extend({token: token}, data), dataType: "json", success: fn});
        })
    }

    return {
        perform: function(action, hash, fn) { get({action: action, hash: hash}, fn) },
        upload: function(url, fn) {get({action: "add-url", s: url}, fn)},
        getTorrents: function(fn) { get({list: 1}, fn) },
        updateTorrents: function(c, fn) { get({list: 1, cid: c}, fn) },
        
    }
}


torrent_info.js
function torrentInfo(torrent) {
    var percent = torrent[4]/10;
    var status = getStatus(torrent[1], percent);
    return {
    hash: torrent[0],
    percent: percent,
    status: status,
    name: getName(torrent[2]),
    size: getSize(torrent[3]),
    downloaded: getSize(torrent[5]),
    uploaded: getSize(torrent[6]),
    ratio: torrent[7]/1000,
    upload_speed_in_bytes: parseInt(torrent[8]),
    upload_speed: getSpeed(torrent[8]),
    download_speed_in_bytes: parseInt(torrent[9]),
    download_speed: getSpeed(torrent[9]),
    eta: (percent == 100) ? "-" : dateHMS(torrent[10]),
    label: torrent[11],
	peers_connected: torrent[12],
	peers_swarm: torrent[13],
	seeds_connected: torrent[14],
	seeds_swarm: torrent[15],
	availability: torrent[16],
	torrent_queue_order: torrent[17],
	remaining: torrent[18],
	actions: getActionWithTitle(status.actions)
   }
}

function getGlobalSpeeds(torrents){
    var speeds = ['download', 'upload'].map(function(type){
	    return getSpeed(torrents.reduce(function(speed, torrent){
            return speed + torrent[type + '_speed_in_bytes'];
            },0))
        })
    return {dl: speeds[0], ul: speeds[1]};  	
}


function getStatus(status, percent) {
	switch (status)
	{
		case 128 : 
		    return { status_text: "check required",
			status_image: "arrow_yellow.png",
			status_bar: "bar_yellow.jpg",
			actions: ["recheck"]}
			
		case 130 : 
			return { status_text: "force checking",
			status_image: "arrow_yellow.png",
			status_bar: "bar_yellow.jpg",
			actions: []}
			
		case 134 :
			return { status_text: "[F]force check'",
			status_image: "arrow_yellow.png",
			status_bar: "bar_yellow.jpg",
			actions: []}
			
		case 136 :
			if (percent == 100)
			{
				return { status_text: "finished",
				status_image: "bar_done.png",
				status_bar: "bar_green.jpg",
				actions: ['forcestart', 'start', 'recheck']}
			}
			else
			{
				return { status_text: "stopped",
				status_image: "arrow_red.png",
				status_bar: "bar_red.jpg",
				actions: ['forcestart', 'start', 'recheck']}
			}
		case 137 :
			if (percent == 100)
			{
				return { status_text: "[F]seeding",
				status_image: "arrow_blueF.png",
				status_bar: "bar_blue.jpg",
				actions: ['stop', 'recheck']}
			}
			else
			{
				return { status_text: "[F]downloading",
				status_image: "arrow_greenF.png",
				status_bar: "bar_green.jpg",
				actions: ['stop', 'recheck']}
			}
			
		case 152 :
			return { status_text: "space missing",
			status_image: "arrow_red.png",
			status_bar: "bar_red.jpg",
			actions: ['recheck']}
    		
		case 162 :
			return { status_text: "checking paused",
			status_image: "arrow_yellow.png",
			status_bar: "bar_yellow.jpg",
			actions: ['unpause']}
    		
		case 194 :
			return { status_text: "checking",
			status_image: "arrow_yellow.png",
			status_bar: "bar_yellow.jpg",
			actions: ['pause']}
			
		case 200 :
			return { status_text: "queued",
			status_image: "arrow_yellow.png",
			status_bar: "bar_yellow.jpg",
			actions: ['forcestart']}
			
		case 201 :
			if (percent == 100)
			{
				return { status_text: "seeding",
				status_image: "arrow_blue.png",
				status_bar: "bar_blue.jpg",
				actions: ['forcestart', 'stop', 'pause', 'recheck']}
			}
			else
			{
				return { status_text: "downloading",
				status_image: "arrow_green.png",
				status_bar: "bar_green.jpg",
				actions: ['forcestart', 'stop', 'pause', 'recheck']}
			}
			
		case 226 :
			return { status_text: "??",
			status_image: "arrow_yellow.png",
			status_bar: "bar_yellow.jpg",
			actions: ['stop', 'recheck']}
			
		case 233 :
			return { status_text: "paused",
			status_image: "arrow_yellow.png",
			status_bar: "bar_yellow.jpg",
			actions: ['forcestart', 'stop', 'unpause', 'recheck']}
			
		default:
			return { status_text: "N/A",
			status_image: "arrow_green.png",
			status_bar: "bar_green.jpg",
			actions: []}
	}
}

function getName(name) {
	return (name.length > 40) ? name.substr(0,37) + "..." : name;
}

function humanFriendlySize(size, types) {
    if (size === 0) return "-";
    var last = types[types.length-1];
    return types.reduce(function(size, type){
        if (isNaN(size)) return size;
        if ((size < 1024) || type == last) return ((size === parseInt(size,10)) ? size : size.toFixed(2)) + type;
        return size/1024;
    },parseInt(size));
}

function getSize(size) {
    return humanFriendlySize(size, [" o", " Ko", " Mo", " Go"]);
}

function getSpeed(size) {
    return humanFriendlySize(size, [" o/s"," Ko/s"," Mo/s"," Go/s"]);
}

function getActionWithTitle(actions) {
    return [['forcestart', 'Force start'], ['start', 'Start'], ['stop', 'Stop'], ['pause', 'Pause'], 
	['unpause', 'Unpause'], ['recheck', 'Force recheck']].filter(function(actionInfo){
		return actions.indexOf(actionInfo[0]) !== -1;
	}).map(function(actionInfo){
	    return {action: actionInfo[0], title: actionInfo[1]}
	})
}

function dateHMS(time)
{
	if (time == 0)
		return "-";
	var addZero = function(v) { return v<10 ? '0' + v : v; };
	var d = new Date(time * 1000); // js fonctionne en milisecondes
	var t = [];
	t.push(addZero(d.getHours())-1);
	t.push(addZero(d.getMinutes()));
	t.push(addZero(d.getSeconds()));
	return t.join(':');
}



js/scripts.js
var $divs;
var $templates = {};
var $utorrent;

$(init);

function refreshRate() {
    if (!isset(localStorage.utorrent_refresh)) localStorage.utorrent_refresh = 2000;
    return localStorage.utorrent_refresh;
}

function init()
{
    $utorrent = setupUtorrent(displayError);
    if ($utorrent.not_configured) return;

    var renderer = create_renderer();

    $divs = {speed: $("#speed"), torrents: $("#torrents"), error: $("#error")};
    
	loadTemplates(function(){
	    renderer.first();
        setInterval(renderer.refresh, refreshRate());
	})
}

function create_renderer() {
    var displayTorrents = function(ts) {
        $divs.error.empty();
        var torrents = [];
        for (key in ts) {
            torrents.push($.extend(ts[key], {action_html: $.jqote($templates.action, ts[key].actions)}))
        }
        
    	if (torrents.length > 0) {
    	    $divs.torrents.jqotesub($templates.torrents, torrents);
            $divs.speed.jqotesub($templates.speed, getGlobalSpeeds(torrents));
        }		
    	else {
    	    $divs.torrents.jqotesub($templates.no_torrents, torrents);
    	}

    	torrents.forEach(attachClickHandlers)
    }
    
    var attachClickHandlers = function(torrent) {
        var torrent_div = $("#"+torrent.hash);
    	var actions = ['start', 'stop'].filter(function(action){
    	    return torrent.status.actions.indexOf(action) !== -1;
    	})

    	actions.forEach(function(action){
    	    torrent_div.find("div.status_image > a, div.status > a").click(function(){
                    perform(torrent, action)
            })
        }) 

        torrent_div.find("div.infos_actions").delegate('a', 'click', function(){
            perform(torrent, $(this).attr("class"))
        })
    };
    
    var perform = function(torrent, action) {
        var prfrm = function(a, fn) {
            $utorrent.perform(a, torrent.hash, fn);
        }
        if (action == 'recheck'){
            prfrm("stop", function(){
                prfrm("recheck");
            });
        } 
        prfrm(action);
    };
    
    var cache_id;
    var torrents;
    
    return {
        first: function() {
            $utorrent.getTorrents(function(result) {
                torrents = result.torrents.reduce(function(ts, data){
                    var t = torrentInfo(data);
                    ts[t.hash] = t;
                    return ts;
                }, {});
                cache_id = result.torrentc;
                displayTorrents(torrents);
            });
        },
        refresh: function(){
           $utorrent.updateTorrents(cache_id, function(result) {
               cache_id = result.torrentc;
               $.each(result.torrentp, function(i, data) {
                   var t = torrentInfo(data);
                   torrents[t.hash] = t;
               })
               $.each(result.torrentm, function(i, data) {
                   delete torrents[data];
               })
               displayTorrents(torrents);
           }) 
        } 
    }
}

function displayError(xhr, error){
    $divs.torrents.empty();
    $divs.speed.empty();
    $divs.error.jqotesub($templates.error);
}

function loadTemplates(fn){
    ['action', 'error', 'torrents', 'no_torrents', 'speed'].reduce(function(acc_fn, template){
        return function() {
            $.get('templates/' + template + '.html.tmpl', function(data){
                $templates[template] = $.jqotec(data);
                acc_fn();
            })
        }
    }, fn)();
}







function getProperties(token, hash)
{
	var xhr = new XMLHttpRequest();
	xhr.open("GET", "http://"+login+":"+password+"@"+ip+":"+port+"/gui/?token="+token+"&action=getprops&hash="+hash, true);
	var requestTimer = setTimeout(function() {
	xhr.abort();
	}, refresh);
	xhr.send();
	xhr.onreadystatechange = function()
	{
		if (xhr.readyState == 4)
		{
			var result = "";
			result = xhr.responseText;
			
			var pos1 = result.indexOf("props");
			pos1 += 10;
			var pos2 = result.indexOf("]");
			pos2 -= 1;
			result = result.slice(pos1, pos2);
			
			var params = result.split(",");
			
			var hash2 = params[0].substr(9, params[0].length-11);
			var trackers = params[1].substr(13, params[1].length-15);
			var ulrate = params[2].substr(10, params[2].length);
			var dlrate = params[3].substr(10, params[3].length);
			var superspeed = params[4].substr(13, params[3].length);
			var dht = params[5].substr(7, params[5].length);
			var pex = params[5].substr(7, params[5].length);
			var seed_override = params[7].substr(17, params[7].length);
			var seed_ratio = params[8].substr(14, params[8].length);
			var seed_time = params[9].substr(13, params[9].length);
			var uslots = params[10].substr(11, params[10].length) ;
			
			result_info = "<div class='infos_titles'>";
			result_info += "up limit: <br/>";
			result_info += "dl limit: <br/>";
			result_info += "seed time: <br/>";
			result_info += "</div>";
			result_info += "<div class='infos_results'>" + ulrate + "<br/>" + dlrate + "<br/>" + seed_time + "<br/>";
			result_info += "</div>";
		}
	}
	return result_info;
}